### PR TITLE
Dcooley issue446

### PR DESCRIPTION
**Pull request by Robinlovelace**

### CODE REVIEW

**Comment by Robinlovelace** [comment]
Building on #738

**Comment by Robinlovelace** [comment]
Heads-up @dcooley this is now at MVP stage in terms of contents that fit in the book. Any comments from you, Jakub, Jannes or anyone welcome!

**Comment by Robinlovelace** [comment]
Idea for a bechmark: round trip `sf` -> `matrix` -> `sf` -> geojson objects. Will be a bit verbose so planning to put in a separate script and just hyperlink to it for now. Long term this needs to be done properly, as per #726.

File: 02-spatial-data.Rmd [hunk 1]
```diff
@@ -670,6 +670,90 @@ st_crs(points_sfc_wgs) # print CRS (only first 4 lines of output shown)
...
+df = data.frame(
+  x = 1:n,
+  y = n:1
+)
+sfheaders::sfg_polygon(obj = df)
+sf = sfheaders::sfg_polygon(obj = df)
+class(sf)
+```
+
+The examples above show the print methods that come with the **sf** library, demonstrating the underlying structures of the objects.
```
Comment (dcooley) [inline 02-spatial-data.Rmd:722]: > show the print methods that come with the **sf** library

I believe this should read "do not show the print methods..."

**Comment by Robinlovelace** [comment]
> I believe this should read "do not show the print methods..."

Good point. Fixed in latest commit. In fact I've taken a slightly different approach now. Interesting that `sfg` objects created by **sf** lack a dimension attribute, while **sfheaders** gives them a dimension. See latest commit. Any other things to say on this?

File: 02-spatial-data.Rmd [hunk 2]
```diff
@@ -695,11 +696,38 @@ The simplest use-case for `**sfheaders**` is demonstrated in the code chunks bel
...
+
+```{r sfheaders-sfg_point, eval=FALSE} 
+v_sfg_sfh # printing without sf loaded
+#>      [,1] [,2]
+#> [1,]    1    1
+#> attr(,"class")
+#> [1] "XY"    "POINT" "sfg" 
 ```
 
+The example above shows the print methods that come with the **sfheaders** package, when **sf** is not loaded, demonstrating the underlying structures of the objects.
```
Comment (dcooley) [inline 02-spatial-data.Rmd:714]: sorry to be pedantic, and it's a minor technicality, but `sfheaders` does not expose a `print` method, so you're seeing the default R print method here.

Comment (Robinlovelace) [reply 02-spatial-data.Rmd:714]: Good to be specific and yes it shows the 'raw' object. Hows the following...

Comment (Robinlovelace) [reply 02-spatial-data.Rmd:714]:
```
The example above shows how `sfg` objects are printed when **sf** is not loaded, demonstrating the underlying structures of the objects. 
```

Comment (dcooley) [reply 02-spatial-data.Rmd]: yeah that's good.

**Comment by Robinlovelace** [comment]
> Another thing that might be worth adding, possibly to the benchmarks, is in {sfheaders} you can CAST between all (valid) types of geometries:

For sure, I did not realise that `sfheaders` had that functionality: amazing!

**Comment by Robinlovelace** [comment]
See casting mentioned in the commit above. This is a great addition to an important package. I think we'll merge it this week, after which you can see how it renders. Any further feedback now or then v. welcome. In the meantime, your approach has me wondering about leveraging the popular and as far as I can tell fast and rock solid Rust crate `geo`. Thoughts with your compiled code and generalisability hat on @dcooley ? This seems future proof to me: https://docs.rs/geo-types/0.1.1/geo_types/enum.Geometry.html#method.as_multipolygon (not tested the casting which is I guess what they mean by `as` - @inicola or georust ppl may be able to clarify) - apologies for tagging people on a somewhat random diversion from the main topic of this PR...

**Comment by Robinlovelace** [comment]
To keep things moving I'm going to merge this, confident this will be a big benefit to readers. Many thanks @dcooley and if you think of anything else for the book let me know. One more question: how best to cite it?

**Comment by lnicola** [comment]
@Robinlovelace sorry, I haven't had a chance yet to figure out what's happening here. Regarding the `as_` methods in `geo`, they aren't C-like casts, but checked ones, returning `None` if the stored geometry wasn't what you asked for.

**Comment by Robinlovelace** [comment]
Thanks for the reply and sorry it's a bit random. Basically, the PR adds mention of an implementation of the simple features specification in C++ and it is fast at casting geometries, by which I mean something like converting an object called `p` with the following geometries

```
POINT (0 0)
POINT (0 1)
POINT (1 1)
POINT (1 0)
```

Into a linestring:

```
LINESTRING (0 0, 0 1, 1 1, 1 0)
```

**Comment by Robinlovelace** [comment]
Heads-up @dcooley and anyone else interested, here is the rendered result of this PR: https://geocompr.robinlovelace.net/spatial-class.html#the-sfheaders-package

**Comment by dcooley** [comment]
There's one typo I found

> For most people users the R function..

should be "For most people _using_ the R function..."

---

> how best to cite it?

I don't honestly mind; I'm happy with `citation("sfheaders")` if that fits in with the others.

**Comment by Robinlovelace** [comment]
Thanks. Post merge fixes applied: https://github.com/Robinlovelace/geocompr/commit/2227cb73f85061e1cd41093ebefc2fba5c06ec1a

File: 02-spatial-data.Rmd [hunk 3]
```diff
@@ -670,6 +670,115 @@ st_crs(points_sfc_wgs) # print CRS (only first 4 lines of output shown)
...
+
+The intention is to allow every **sfheaders** operation implemented in R to have a C++ implementation, as described in the [`Cpp` vignette](https://dcooley.github.io/sfheaders/articles/Cpp.html).
+For most people users the R functions will be more than sufficient to benefit from the computational speed of the package.
+**sfheaders** was developed separately from **sf**, but aims to be fully compatible, creating valid `sf` objects of the type described in preceding sections.
+
+The simplest use-case for `**sfheaders**` is demonstrated in the code chunks below with examples of building  `sfg`, `sfc` and `sf` objects showing:
+
+- a vector converted to `sfg_POINT`
+- a matrix converted to `sfg_LINESTRING`
+- a data.frame converted to `sfg_POLYGON`
```
Comment (jannes-m) [inline 02-spatial-data.Rmd:697]: write data frame

File: 02-spatial-data.Rmd [hunk 4]
```diff
@@ -670,6 +670,115 @@ st_crs(points_sfc_wgs) # print CRS (only first 4 lines of output shown)
...
+
+```{r, echo=FALSE, eval=FALSE}
+# (although `sfg` objects created with **sfheaders** have a dimension while `sfg` objects created with the **sf** package do not)
+waldo::compare(v_sfg_sf, v_sfg_sfh)
+dim(v_sfg_sf)
+dim(v_sfg_sfh)
+attr(v_sfg_sfh, "dim")
+```
+
+The next examples shows how **sfheaders** creates objects creates `sfg` objects from matrices:
```
Comment (jannes-m) [inline 02-spatial-data.Rmd:734]: delete first instance of creates objects

File: 02-spatial-data.Rmd [hunk 5]
```diff
@@ -670,6 +670,115 @@ st_crs(points_sfc_wgs) # print CRS (only first 4 lines of output shown)
...
+In each of these examples the CRS (coordinate reference system) is not defined.
+If you plan on doing any calculations or geometric operations using **sf** functions, set the CRS as follows (see Chapter \@ref(reproj-geo-data) for details):
+
+```{r sfheaders-crs}
+df_sf = sfheaders::sf_polygon(obj = df)
+sf::st_crs(df_sf) = 4326
+```
+
+**sfheaders** is also good at 'deconstructing' and 'reconstructing' `sf` objects, meaning converting geometry columns into data frames that contain data on the coordinates of each vertex and geometry feature (and multi-feature) ids.
+It is fast and reliable at 'casting' geometry columns to different types, a topic covered in Chapter \@ref(eometric-operations).
```
Comment (jannes-m) [inline 02-spatial-data.Rmd:779]: write \@ref(geometric-operations)

**Review by jannes-m — COMMENTED** [review]
Excellent addition to the book!!!

File: 02-spatial-data.Rmd (hunk 4 above)
Comment (Robinlovelace) [reply 02-spatial-data.Rmd:734]: :+1:

**Pull request merged**